File: 0090-subsets-ii/0090-subsets-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
function subset(index, nums, subArr, result) {

    if ( index >= nums.length ) {
        result.push([...subArr]);
        return;
    }

    subArr.push(nums[index])
    subset(index+1, nums, subArr, result)
    subArr.pop()

    while (index + 1 < nums.length && nums[index] === nums[index + 1]) {
        index++;
    }
    
    subset(index+1, nums, subArr, result)

    return result;

}
var subsetsWithDup = function(nums) {
    nums.sort((a, b) => a - b);
    let subArr = [];
    let result = [];
    return subset(0, nums, subArr, result);
    
};

File: 0090-subsets-ii/0090-subsets-ii.java
class Solution {
    private void generateSubsets(int[] nums, int start, List<List<Integer>> result, List<Integer> subResult) {
        result.add(new ArrayList<>(subResult));
        for ( int i=start; i<nums.length; i++ ) {
            if ( i>start && nums[i] == nums[i-1] ) continue;
            subResult.add(nums[i]);
            generateSubsets(nums, i+1, result, subResult);
            subResult.remove(subResult.size()-1);
        }
    }
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> subResult = new ArrayList<>();
        Arrays.sort(nums);
        generateSubsets(nums, 0, result, subResult);
        return result;
    }
}